Implement basic authentication for Docker registries

flx42 opened the pull request:
There are a few problems currently:
- The Docker registry to use cannot be modified for `singularity import`
- The `--no-token` option is tricky. A token might be necessary for a registry, even if you don't need an account. We should attempt without a token, then try to get a token if we receive a 401.
- Basic authentication is not supported, so I don't think there is a way to fetch private repos.
- Token caching might not be a good idea: the token might expire when downloading large images over a slow connection. I think it's safer to get a new token per layer, and it makes the code simpler imo.

To specify the registry for `singularity import`:
```
export SINGULARITY_DOCKER_REGISTRY='--registry myrepo'
export SINGULARITY_DOCKER_AUTH='--username fabecassis --password [mygreatpassword]'
```
Having to export the password as an environment variable might not be great, I'm open to better suggestions. Maybe an authentication file like Docker does.

I tested my patch with “singularity import” on the following:
- DockerHub: a public and a private repo
- Google Container Registry: gcr.io/tensorflow (public)
- Githost.io (GitLab): a private repo 

It seemed to work, please test!

vsoch commented on libexec/bootstrap/modules-v2/build-docker.sh:
```diff
@@ -85,19 +85,12 @@ fi
-
+SINGULARITY_DOCKER_USERNAME=`singularity_key_get "Username" "$SINGULARITY_BUILDDEF"`
+SINGULARITY_DOCKER_PASSWORD=`singularity_key_get "Password" "$SINGULARITY_BUILDDEF"`
+if [ -n "${SINGULARITY_DOCKER_USERNAME:-}" ] && [ -n "${SINGULARITY_DOCKER_PASSWORD:-}" ]; then
```
I like these keys `Username` and `Password` as they could be generally used for different bootstrap options. @gmkurtzer - is there an easy way to make it so capitalization doesn't matter? USERNAME should work equivalently to username or Username.

gmkurtzer replied:
They shouldn't be capitalization specific already:

```
    if OUT=`egrep -i "^$KEY:" $FILE`; then
        echo "$OUT" | head -n 1 | sed -e "s@^$KEY:\s*@@i" | sed -e "s@\s*#.*@@"
        return 0
    fi
```

vsoch replied:
ok, awesome. The last time I tried was a long time ago, I remember specifying from instead of From didn't work, glad that's fixed!

vsoch commented on libexec/helpers/image-import.sh:
```diff
@@ -54,7 +54,7 @@ fi
         CONTAINER_NAME=`echo "$IMPORT_URI" | sed -e 's@^docker://@@'`
-        SINGULARITY_IMPORT_GET="$SINGULARITY_libexecdir/singularity/python/cli.py --rootfs '$SINGULARITY_ROOTFS' --docker '$CONTAINER_NAME' --cmd"
+        SINGULARITY_IMPORT_GET="$SINGULARITY_libexecdir/singularity/python/cli.py --rootfs '$SINGULARITY_ROOTFS' --docker '$CONTAINER_NAME' ${SINGULARITY_DOCKER_REGISTRY:-} ${SINGULARITY_DOCKER_AUTH:-} --cmd"
     ;;
```
I didn't look at this part when @gmkurtzer implemented - it looks like it doesn't include logic for including / not including the command as runscript. Right now, the default (without `--cmd`) is to use the entrypoint, which is the more proper of the two (CMD and ENTRYPOINT) as it is the container's executable. Adding the `--cmd` argument overrides this default, and changes the runscript to be CMD. For import, we would likely want the default to be the same (not having CMD) and akin to the docker credentials, have it be an envrionmental variable. Thoughts @gmkurtzer @flx42 ?

gmkurtzer replied:
Ahh, good catch. Probably best to leave it off then.

vsoch replied:
Agreed - we can stick to default (entrypoint) for now, and then if someone says "heeeey I want CMD on import too!" then it should be pretty easy to add.

flx42 replied:
It should be a different patch set anyway, so I'll leave it this way :)

vsoch replied:
Would you mind removing the --cmd bit at the end? If it's a hassle I'd be glad to submit a PR to your PR, haha. I think it's important to have this behavior be consistent asap between bootstrap and import (which it currently isn't).

vsoch commented on libexec/python/cli.py:
```diff
@@ -112,11 +115,12 @@ def main():
-    logger.info("Do registry authentication set to %s", doauth)
+    # Does the registry require authentication?
+    auth = None
+    if args.username is not None and args.password is not None:
```
Can we move this guy into a function in docker/utils.py that takes a username,password, and returns an updated auth object in the same fashion? This will be useful in case we need to sneak it in other places at some point - we'd want them all to change based on pinging the same function.

flx42 replied:
Done

vsoch commented on libexec/bootstrap/modules-v2/build-docker.sh:
```diff
@@ -85,19 +85,12 @@ fi
+    message 1 "Username: $SINGULARITY_DOCKER_USERNAME\n"
+    message 1 "Password: [hidden]\n"
+    SINGULARITY_DOCKER_AUTH="--username $SINGULARITY_DOCKER_USERNAME --password $SINGULARITY_DOCKER_PASSWORD"
 else
```
I like this :)

vsoch commented on libexec/python/cli.py:
```diff
@@ -185,12 +189,6 @@ def main():
                                disable_cache = args.disable_cache)
 
         layers = []
-
```
I have mixed feelings about this because it seems redundant, but you are correct that a large or timely download could mean an invalid token. It will probably be annoying from a user perspective to have the token printed each time, a fix could be to obtain it each time (but only have the message appear in the debug logs).

flx42 replied:
Actually, I removed your `logger.info("Obtaining token: %s", base)` line.
A line will only be printed if a token is **not** required: `logger.info("No token required for registry %s", registry)`. But we can remove it, no need to be overly verbose if everything is fine...

vsoch replied:
Excellent. Yes, this line only makes sense if we did it once.

vsoch commented on libexec/python/docker/api.py:
```diff
@@ -47,24 +52,46 @@ def create_runscript(cmd,base_dir):
 
-def get_token(repo_name,namespace="library",scope="repository",permission="pull"):
+def get_token(registry,repo_name,namespace,scope="repository",permission="pull",auth=None):
     '''get_token will use version 2.0 of Docker's service to return a token with given permission and scope - this
```
What is the reason for requiring the registry here? Akin to the other functions, why not have the default as None, and then use api_base if it is? We might want to turn those three lines into a proper function.

flx42 replied:
I think it makes less sense here, I think functions like `get_config`, `get_manifest` or `get_layer` are part of the actual public API of your module. But it would be weird to have `get_token` part of the public API, you always need a token for something :)
But I can change it for the sake of consistency if you want.

vsoch replied:
Do you mean having get_token be provided at all? For organization - I don't see any reason that someone might want to use it, you are right that it might be more appropriate as a utility or authentication function - if we had a larger code base it would be appropriate to have an auth.py in the docker folder, but I didn't think this level of organization was necessary (yet anyway).  For what my comment was about - Is there a reason for not having the registry default to index.docker.io as we do in the other functions? That should be easy to add, even it's not a complete function (eg, the same three lines of code are repeated in multiple places in the file!)

flx42 replied:
Yeah, I believe `get_token` is more an utility function. It should only be called from `get_config`, `get_manifest`, `get_layer`, etc. And these functions can indeed default to `index.docker.io`, but no need to do it twice IMO.
But yeah, it's easy to fix :)

vsoch replied:
cool, thanks!

flx42 replied:
So I should add the default?

vsoch replied:
Yes please!

vsoch commented on libexec/python/docker/api.py:
```diff
@@ -47,24 +52,46 @@ def create_runscript(cmd,base_dir):
-def get_token(repo_name,namespace="library",scope="repository",permission="pull"):
+def get_token(registry,repo_name,namespace,scope="repository",permission="pull",auth=None):
     '''get_token will use version 2.0 of Docker's service to return a token with given permission and scope - this
     function does work, but the token doesn't seem to work when used with other functions below for authentication
```
oups, need to update this line since we fixed the "doesn't work" part :P

flx42 replied:
Ah yes, correct, the doc needs updating. Actually parameters `scope` and `permission` are gone. We shouldn't need to pass these values to a function, they should be inferred from the HTTP response after we get a `401`.

vsoch replied:
+1. Also - I'll take care of updating the online docs to give details about using the command line token for import once this is merged. I'll ping you with the link to check over for a second set of eyes.

vsoch commented on libexec/python/docker/api.py:
```diff
@@ -157,11 +184,11 @@ def get_manifest(repo_name,namespace,repo_tag="latest",registry=None,auth=True):
-        token = get_token(repo_name=repo_name,
-                          namespace=namespace,
-                          permission="pull")
+    token = get_token(registry=registry,
```
I like this logic much better.

flx42 committed:
Implement basic authentication for Docker registries

The --no-token option has been removed, we try to detect if a token is needed or not.
Token caching has been removed: the token might expire when downloading large images over a slow connection.

Signed-off-by: Felix Abecassis <fabecassis@nvidia.com>
